flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala

```diff
@@ -0,0 +1,244 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.calcite
+
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo._
+import org.apache.flink.api.common.typeinfo._
+import org.apache.flink.api.java.typeutils.ValueTypeInfo._
+import org.apache.flink.table.`type`.DecimalType
+import org.apache.flink.table.api.TableException
+import org.apache.flink.table.calcite.FlinkTypeFactory.typeInfoToSqlTypeName
+import org.apache.flink.table.typeutils._
+
+import org.apache.calcite.avatica.util.TimeUnit
+import org.apache.calcite.jdbc.JavaTypeFactoryImpl
+import org.apache.calcite.rel.`type`._
+import org.apache.calcite.sql.SqlIntervalQualifier
+import org.apache.calcite.sql.`type`.SqlTypeName
+import org.apache.calcite.sql.`type`.SqlTypeName._
+import org.apache.calcite.sql.parser.SqlParserPos
+
+import java.util
+
+import scala.collection.JavaConverters._
+
+/**
+  * Flink specific type factory that represents the interface between Flink's [[TypeInformation]]
```

Comment:
We need add a TODO or jira to track all the convert between `TypeInformation` and Calcite's `RelDataType`. From my understanding, in most cases, we only need to convert between `InternalType` and `RelDataType`, and the usage of `TypeInformation` should verified case by case.

Reply:
Currently, `FlinkTypeFactory` only contains very basic features based on `TypeInformation` to make this PR work. @JingsongLi have created two jiras: https://issues.apache.org/jira/browse/FLINK-11803 and https://issues.apache.org/jira/browse/FLINK-11802 to improve this.